src/Collector/Finder/VariableFinder.php

```diff
@@ -18,24 +19,33 @@
 final class VariableFinder
 {
     /**
-     * @var array<string, array<string, Variable[]>>
+     * @var array<string, array<string, array<string, Variable>>>
      */
-    private array $collectedVariables = [];
+    private array $assignedVariables = [];
+
+    /**
+     * @var array<string, array<string, array<string, Variable>>>
+     */
+    private array $declaredVariables = [];
 
     private MethodCallFinder $methodCallFinder;
 
     public function __construct(CollectedDataNode $collectedDataNode, TypeSerializer $typeSerializer, MethodCallFinder $methodCallFinder)
     {
         $this->methodCallFinder = $methodCallFinder;
 
-        $collectedVariables = VariableCollector::loadData($collectedDataNode, $typeSerializer, CollectedVariable::class);
-        foreach ($collectedVariables as $collectedVariable) {
-            $className = $collectedVariable->getClassName();
-            $methodName = $collectedVariable->getMethodName();
-            if (!isset($this->collectedVariables[$className][$methodName])) {
-                $this->collectedVariables[$className][$methodName] = [];
-            }
-            $this->collectedVariables[$className][$methodName][] = $collectedVariable->getVariable();
+        $assignedVariables = VariableCollector::loadData($collectedDataNode, $typeSerializer, CollectedVariable::class);
+        foreach ($assignedVariables as $variable) {
+            $className = $variable->getClassName();
+            $methodName = $variable->getMethodName();
+            $this->assignedVariables[$className][$methodName] = Variable::union($this->assignedVariables[$className][$methodName] ?? [], [$variable->getVariable()]);
```

Comment:
Union on variable should just union actual type of variable with new type. Not create new array of all variables. This should be service / helper outside of Variable

Reply:
I was thinking about helper too. I will change it.

Reply:
Any idea of good name? Simply VariableHelper?

Reply:
probably yes :)